Documentation: JWT Key Rotation (#849)
raft-tech/TANF-app: docs/Technical-Documentation/jwt-key-rotation.md

# Rotating JWT Keys

## Context

To maintain good security, we will periodically rotate our JWT keys. This document outlines the process of how to do this.

> **Warning** Production sites will need to be taken down for maintenance when rotating keys, as the rotation will automatically invalidate all current sessions.

Author: I definitely think a "maintenance mode" would be the way to do it, but we don't currently have anything planned for that so I thought including it here was a bit of a stretch. I think this does call to our attention the need to plan for that, though.

> **Warning** Production sites will need to be taken down for maintenance when rotating keys, as the rotation will automatically invalidate all current sessions.

Reviewer: Agreed, I'd think a separate issue to do that after ATO would be a good call.

### 1. Generate New Keys

In your Mac terminal (or bash terminal in Windows), enter the following command:
```bash=
ssh-keygen -t rsa -b 4096 -m PEM -f jwtRS256.key
```
You will receive the following response. Do not add a passphrase when prompted.
```
Enter passphrase (empty for no passphrase):
Enter same passphrase again:
Your identification has been saved in jwtRS256.key
Your public key has been saved in jwtRS256.key.pub
The key fingerprint is:
# returns key fingerprint
The key's randmart image is:
# returns key's randmart image
```
Then enter
```bash=
openssl rsa -in jwtRS256.key -pubout -outform PEM -out jwtRS256.key.pub
```

You can now check the contents of your keys with these commands
```bash=
cat jwtRS256.key
# returns private key
cat jwtRS256.key.pub
# returns public key
```

### 2. Base64 Encode Private Key

We use Base64 Encoded Private Keys to make it easier to save to cloud environments and local `.env` files.

```bash=

> openssl enc -base64 -in jwtRS256.key -out jwtRS256.key.base64

Reviewer: If you don't need to actually write this file we can convert it on the fly with:
```
base64 -w 0 jwtRS256.key.pub
```

Additionally, by using the `-w 0` flag this command will output the entire value on a single line - which can help prevent errors when copying that value over and manually removing the spaces.

Author: I thought it was useful to save it just in case of any copy/paste errors we would have a local copy. Good call on the `-w 0` flag.

cat jwtRS256.key.base64
```

### 3. Copy Keys

#### Dev/Staging Environments
1. Distribute the private key to development staff securely to copy to `.env` files as the value for key `JWT_KEY`

> 2. Update the variable `JWT_KEY` with the private key in cloud.gov backend development and staging environments

Reviewer: ```suggestion
2. Update the environment variable `JWT_KEY` with the private key in cloud.gov backend development and staging environments
```

> 3. Login to the [Login.gov Sandbox](https://dashboard.int.identitysandbox.gov/) and update the public key there

Reviewer: Would it be safe to put a screenshot of where to go in their UI here? As long as nothing sensitive is shown I think that would be helpful to have here.

Reviewer: Also, does Login.gov expect the key in PEM format or Base 64?

Author: PEM format. I have updated the doc with a note and a screenshot

#### CI/CD Environment
**Note** _Please generate a separate set of keys for the CI/CD environment_
1. Distribute the private key to development staff securely to copy to `.env` files as the value for key `JWT_CERT_TEST`

> 2. Update the variable `JWT_KEY_TEST` in CircleCI with the new public key.

Reviewer: I see we have both `JWT_CERT_TEST` and `JWT_KEY_TEST` saved in Circle CI currently but I don't see a reference to `JWT_KEY_TEST` anywhere in the code base - are we supposed to have both?

Author: I think we need `JWT_KEY_TEST` in CircleCI and `JWT_CERT_TEST` defined in our local `.env` files to send to CircleCI. I did notice we have some things defined in different places that we don't need, so a follow up issue to clean those up would probably be in order.

Reviewer: Looks like we have a stub ticket we can use to track this: https://github.com/raft-tech/TANF-app/issues/597

# Rotating JWT Keys

## Context

> To maintain good security, we will periodically rotate our JWT keys. This document outlines the process of how to do this.

Reviewer: ```suggestion
To maintain good security, we will periodically rotate the JWT keys used to control authentication and authorization to our application. This document outlines the process of how to do this.
```

**Warning** Production sites will need to be taken down for maintenance when rotating keys, as the rotation will automatically invalidate all current sessions.

### 1. Generate New Keys

In your Mac terminal (or bash terminal in Windows), enter the following command:
```bash=
ssh-keygen -t rsa -b 4096 -m PEM -f jwtRS256.key
```
You will receive the following response. Do not add a passphrase when prompted.
```
Enter passphrase (empty for no passphrase):
Enter same passphrase again:
Your identification has been saved in jwtRS256.key
Your public key has been saved in jwtRS256.key.pub
The key fingerprint is:
# returns key fingerprint
The key's randmart image is:
# returns key's randmart image
```
Then enter
```bash=
openssl rsa -in jwtRS256.key -pubout -outform PEM -out jwtRS256.key.pub
```

You can now check the contents of your keys with these commands
```bash=
cat jwtRS256.key
# returns private key
cat jwtRS256.key.pub
# returns public key
```

### 2. Base64 Encode Private Key

We use Base64 Encoded Private Keys to make it easier to save to cloud environments and local `.env` files.

```bash=

> openssl enc -base64 -w 0 -in jwtRS256.key -out jwtRS256.key.base64

Reviewer 2: @jtwillis92 what does `-w 0` mean? this returned an error on my end and `openssl enc -help` didnt shed any light. i removed this from the script and doesnt return any errors, but will this cause any problems for us?

Reviewer 2: thanks @jtwillis92 can you update the doc w/ this distinction? I'll proceed with rotating and finishing up the qasp review.

# Rotating JWT Keys

## Context

To maintain good security, we will periodically rotate the JWT keys used to control authentication and authorization to our application. This document outlines the process of how to do this.

**Warning** Production sites will need to be taken down for maintenance when rotating keys, as the rotation will automatically invalidate all current sessions.

### 1. Generate New Keys

In your Mac terminal (or bash terminal in Windows), enter the following command:
```bash=
ssh-keygen -t rsa -b 4096 -m PEM -f jwtRS256.key
```

> You will receive the following response. Do not add a passphrase when prompted.

Reviewer 2: see above suggested change that would replace this portion of output if acceptable.

Author: I'm not sure which portion this refers to. Are there specific line numbers?

Reviewer 2: 15-28 (as you noted below)

```
Enter passphrase (empty for no passphrase):
Enter same passphrase again:
Your identification has been saved in jwtRS256.key
Your public key has been saved in jwtRS256.key.pub
The key fingerprint is:
# returns key fingerprint
The key's randmart image is:
# returns key's randmart image
```
Then enter
```bash=

> openssl rsa -in jwtRS256.key -pubout -outform PEM -out jwtRS256.key.pub

Author: @ADPennington This looks like we could remove lines 15-28 altogether. Is that correct?

> openssl rsa -in jwtRS256.key -pubout -outform PEM -out jwtRS256.key.pub

Reviewer 2: yep! 15-28 would be replaced with something like: 

You will receive the following response. Answering questions when prompted is not necessary.

```
-----
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Country Name (2 letter code) [AU]:
State or Province Name (full name) [Some-State]:
Locality Name (eg, city) []:
Organization Name (eg, company) [Internet Widgits Pty Ltd]:
Organizational Unit Name (eg, section) []:
Common Name (e.g. server FQDN or YOUR name) []:
Email Address []:
```

```

You can now check the contents of your keys with these commands
```bash=
cat jwtRS256.key
# returns private key

> cat jwtRS256.key.pub

Reviewer 2: ```suggestion
cat jwtRS256pub.crt
```

# returns public key
```

### 2. Base64 Encode Private Key

We use Base64 Encoded Private Keys to make it easier to save to cloud environments and local `.env` files.

```bash

> openssl enc -base64 -in jwtRS256.key -out jwtRS256.key.base64

Reviewer 2: ```suggestion
openssl enc -base64 -in jwtRS256prv.pem -out jwtRS256prv.pem.base64

```

> cat jwtRS256.key.base64

Reviewer 2: ```suggestion
cat jwtRS256prv.pem.base64
```

```

> NOTE: Linux users must disable line wrapping by adding the argument `-w 0` to get a properly formatted one-line value.

Reviewer 2: not sure if we'd still need this if suggestion above is acceptable.

Author: I think this would still be applicable to Linux users

Reviewer 2: good catch! would line 51 be replaced with a modified version of  line 13 for linux users?

Author: I don't believe it would be necessary to update line 13, since that one isn't base64

```bash
-----
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Country Name (2 letter code) [AU]:
State or Province Name (full name) [Some-State]:
Locality Name (eg, city) []:
Organization Name (eg, company) [Internet Widgits Pty Ltd]:
Organizational Unit Name (eg, section) []:
Common Name (e.g. server FQDN or YOUR name) []:
Email Address []:
```

You can now check the contents of your keys with these commands
```bash=
cat jwtRS256prv.pem
# returns private key
cat jwtRS256pub.crt
# returns public key
```

### 2. Base64 Encode Private Key

We use Base64 Encoded Private Keys to make it easier to save to cloud environments and local `.env` files.

```bash
openssl enc -base64 -in jwtRS256prv.pem -out jwtRS256prv.pem.base64

cat jwtRS256prv.pem.base64
```

NOTE: Linux users must disable line wrapping by adding the argument `-w 0` to get a properly formatted one-line value.
```bash

> openssl enc -base64 -w 0 -in jwtRS256.key -out jwtRS256.key.base64

Reviewer 2: ```suggestion
openssl enc -base64 -w 0 -in jwtRS256prv.pem -out jwtRS256prv.pem.base64
```

> cat jwtRS256.key.base64

Reviewer 2: ```suggestion
cat jwtRS256prv.pem.base64
```

```

### 3. Copy Keys

#### Dev/Staging Environments
1. Distribute the private key to development staff securely to copy to `.env` files as the value for key `JWT_KEY`
2. Update the environment variable `JWT_KEY` with the private key in cloud.gov backend development and staging environments
3. Login to the [Login.gov Sandbox](https://dashboard.int.identitysandbox.gov/) and update the public key there

> Note: Login.gov requires the key to be uploaded in PEM format, which is the format we produced in the `jwtRS256.key.pub` file.

Reviewer 2: ```suggestion
Note: Login.gov requires the key to be uploaded in PEM format, which is the format we produced in the `jwtRS256pub.crt` file.
```
